### PR TITLE
provider: added k3d provider and node lifecycle handlers

### PR DESCRIPTION
**Pull request by harshanarayana**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

This PR enables a new `E2EClusterProvider` for `k3d` based infra. 

    This change includes the following changes and features.

1. Added a new Interface type `E2EClusterProviderWithImageLoaderAndNodeLifecycle` which can be used to setup providers that extend the cluster lifecycle function around the nodes.
2. Enabled a `k3d` based provider with support for Node lifecycle management.
3. Existing Image loader related function and interfaces were augmented with `args ...string` to be able to provide additional arguments in case if the image load handlers need some of the additional config.

#### Which issue(s) this PR fixes:
Fixes #

#### Special notes for your reviewer:
I have also extended the `E2EClusterProviderWithImageLoader` 's Image load related methods to take an additional `arg ...string` value to account for additional options that one can pass while performing image load operation alone. 

This comes in handy with `k3d` where it provides a few different mode of loading images and cleanup workflows. 

This change also doesn't break any existing API contract extended by the `envfuncs`.

#### Does this PR introduce a user-facing change?

```release-note
enabled a new `E2eClusterProvider` for `k3d` that can be used as a new cluster provider for running e2e tests.
```

#### Additional documentation e.g., Usage docs, etc.:
```docs
NA
```

### CODE REVIEW

**Comment by harshanarayana** [comment]
`go install` of k3d is failing due to an issue similar to https://github.com/k3d-io/k3d/issues/1324

**Comment by harshanarayana** [comment]
/hold 

Holding this under the `k3d` install issue is resolved.

File: support/utils/command.go [hunk 1]
```diff
@@ -34,41 +35,57 @@ var commandRunner = gexe.New()
...
 	log.V(4).InfoS("Installing provider tooling using go install", "command", installCommand)
-	p := commandRunner.RunProc(installCommand)
-	if p.Err() != nil {
-		return "", fmt.Errorf("failed to install %s: %s", pPath, p.Err())
+	p := commandRunner.NewProc(installCommand)
+	p.SetStdout(&stdout)
+	p.SetStderr(&stderr)
+	result := p.Run()
+	if result.Err() != nil {
+		return "", fmt.Errorf("failed to install %s: %s: \n %s", pPath, result.Result(), stderr.String())
```
Comment (harshanarayana) [inline support/utils/command.go:51]: Adding these to be able to capture and log the install failures better during the provider install.

**Review by cpanato — APPROVED** [review]
/lgtm
thanks

**Comment by harshanarayana** [comment]
Holding this until @cpanato  @vladimirvivien  @ShwethaKumbla also takes look

**Review by cpanato — COMMENTED** [review]
/lgtm

File: support/types.go [hunk 2]
```diff
@@ -94,12 +114,48 @@ type E2EClusterProviderWithImageLoader interface {
...
+
+// E2EClusterProviderWithImageLoaderAndNodeLifecycle is an interface that extends the E2EClusterProviderWithImageLoader
+// interface to provide a mechanism to add/remove nodes from the cluster as part of the E2E Test workflow.
+//
+// This can be useful while performing the e2e test that revolves around the node lifecycle events.
+// eg: You have a kubernetes controller that acts upon the v1.Node resource of the k8s and you want to
+// test out how the Remove operation impacts your workflow.
+// Or you want to simulate a case where one or more node of your cluster is down and you want to see how
+// your application reacts to such failure events.
+type E2EClusterProviderWithImageLoaderAndNodeLifecycle interface {
```
Comment (vladimirvivien) [inline support/types.go:135]: Oof, that is a long name.
Let's reconsider making it simpler or break into multiple (embedded) interfaces.
* `E2EClusterProvier`
* `E2EClusterProviderWithImgLoader`
*`E2EClusterProviderWithLifeCycle`

**Review by vladimirvivien — COMMENTED** [review]
@harshanarayana this a great (massive) PR.
Thanks for doing this and I am sure the community will appreciate the k3d integration.
I left some comments and we can work on getting this to the finish line.

File: pkg/stepfuncs/nodelifecycle_funcs.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+	"sigs.k8s.io/e2e-framework/pkg/types"
+	"sigs.k8s.io/e2e-framework/support"
+)
+
+// PerformNodeOperation returns a step function that performs a node operation on a cluster.
+// This can be integrated as a setup function for a feature in question before the feature
+// is tested.
+func PerformNodeOperation(clusterName string, action support.NodeOperation, node *support.Node, args ...string) types.StepFunc {
```
[the remark this answers is not in the document]
Comment (harshanarayana) [reply pkg/stepfuncs/nodelifecycle_funcs.go:31]: Ack

File: support/k3d/k3d.go [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+
+func (c *Cluster) SetDefaults() support.E2EClusterProvider {
+	if c.path == "" {
+		c.path = "k3d"
+	}
+	return c
+}
+
+func (c *Cluster) WaitForControlPlane(ctx context.Context, client klient.Client) error {
+	log.V(4).Info("k3d doesn't implement a WaitForControlPlane handler. The --wait argument passed to the `kwokctl` should take care of this already")
```
[the remark this answers is not in the document]
Comment (harshanarayana) [reply support/k3d/k3d.go:294]: 😞 That was a copy paste mistake. Fixing it.

File: support/types.go (hunk 2 above)
Comment (harshanarayana) [reply support/types.go:135]: Ack. using `E2EClusterProviderWithLifeCycle` instead

Comment (harshanarayana) [reply support/types.go:135]: @vladimirvivien  Do you want to check the image loader related interface name too ?

**Comment by harshanarayana** [comment]
@vladimirvivien  PTAL. I has pushed the requested changes.

**Review by ShwethaKumbla — COMMENTED** [review]
LGTM

File: support/k3d/k3d.go [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+}
+
+func (c *Cluster) StopNode(ctx context.Context, node *support.Node, args ...string) error {
+	cmd := fmt.Sprintf("%s node stop %s", c.path, node.Name)
+	if len(args) > 0 {
+		cmd = fmt.Sprintf("%s %s", cmd, strings.Join(args, " "))
+	}
+	log.V(4).InfoS("Stopping node in k3d cluster", "command", cmd)
+	p, stdout, stderr := utils.FetchSeperatedCommandOutput(cmd)
+	if p.Err() != nil || (p.Exited() && p.ExitCode() != 0) {
```
Comment (vladimirvivien) [inline support/k3d/k3d.go:371]: Oh boy, you are pushing that package (gexe) to it's limit. Love it 😉

File: pkg/utils/nodelifecycle.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+import (
+	"context"
+	"fmt"
+
+	"sigs.k8s.io/e2e-framework/support"
+)
+
+// PerformNodeLifecycleOperation performs a node operation on a cluster. These operations can range from Add/Remove/Start/Stop.
+// This helper is re-used in both node lifecycle handler used as types.StepFunc or env.Func
+func PerformNodeLifecycleOperation(ctx context.Context, action support.NodeOperation, node *support.Node, args ...string) error {
```
Comment (vladimirvivien) [inline pkg/utils/nodelifecycle.go:28]: Does this function only applies when using K3d or do you think orther providers will use that. 
If only k3d, maybe it belongs in third_party package path if that is the case.

File: support/k3d/k3d.go [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+
+func (c *Cluster) SetDefaults() support.E2EClusterProvider {
+	if c.path == "" {
+		c.path = "k3d"
+	}
+	return c
+}
+
+func (c *Cluster) WaitForControlPlane(ctx context.Context, client klient.Client) error {
+	log.V(4).Info("k3d doesn't implement a WaitForControlPlane handler. The --wait argument passed to the `k3d` should take care of this already")
```
Comment (vladimirvivien) [inline support/k3d/k3d.go:294]: I would reword this and just say that "the k3d provider doesn't implement a WaitForControlPlane as the provider automatically wait for the control plane"

Comment (vladimirvivien) [reply support/k3d/k3d.go:294]: One more question, what --wait argument ?
Are you saying that you can pass a --wait argument during cluster create ?

**Review by vladimirvivien — COMMENTED** [review]
Making progress @harshanarayana 
Added some more comments.

File: pkg/utils/nodelifecycle.go (hunk 6 above)
Comment (harshanarayana) [reply pkg/utils/nodelifecycle.go:28]: Other providers like mini kube can support this.. And `kwok` technically can be simulated to enable node add and remove workflows. So it can work with other providers as well. We can do that even for kind. But we will have to use adhoc docker command to simulate power on and off kind of workflows

File: support/k3d/k3d.go (hunk 7 above)
Comment (harshanarayana) [reply support/k3d/k3d.go:294]: Yes.. K3d has an arg `--wait` with cluster create.. Which also defaults to true anyway.. Let me reword the message as you suggested.

**Comment by harshanarayana** [comment]
@vladimirvivien  PTAL when you can. Addressed the required changes in the message for wait for control plane.

File: support/k3d/k3d.go [hunk 8]
```diff
@@ -0,0 +1,399 @@
+/*
```
Comment (vladimirvivien) [inline support/k3d/k3d.go:1]: This should be in third-party.
I see that kwok in that package, but it should have also been in third-party.
package support is a place for API definitions for third-party support.
`kind` was supposed to be an example implementation for others (maybe it should have been in third-party as well)

File: support/utils/command.go [hunk 9]
```diff
@@ -34,41 +35,57 @@ var commandRunner = gexe.New()
...
+
+	if providerPath := gexe.ProgAvail(provider); providerPath != "" {
 		log.V(4).Infof("Installed %s at %s", pPath, providerPath)
 		return provider, nil
 	}
 
-	p = commandRunner.RunProc("ls $GOPATH/bin")
-	if p.Err() != nil {
-		return "", fmt.Errorf("failed to install %s: %s", pPath, p.Err())
+	p = commandRunner.NewProc("ls $GOPATH/bin")
```
Comment (vladimirvivien) [inline support/utils/command.go:65]: Do you think `gexe` could be improved to make output caputure easier ? If so, please open an issue.

**Review by vladimirvivien — COMMENTED** [review]
More comments.

File: support/k3d/k3d.go (hunk 8 above)
Comment (harshanarayana) [reply support/k3d/k3d.go:1]: Good point.. I can move them all. But we have to retain some sort of aliasing to aid the import backward compatibility.. At least for kwok and the kind that we support today in an available release.. Right? 


If you are ok with that, I will take care of the required package refactoring

Comment (vladimirvivien) [reply support/k3d/k3d.go:1]: Hi @harshanarayana if type aliasing will help maintain backward compat, then do it. 
At the very least, we should just move kwok.  Thanks for doing it.

**Comment by vladimirvivien** [comment]
@harshanarayana is this still on your radar for 2025?

**Comment by harshanarayana** [comment]
I will get this done before we cut the christmas release..

**Comment by harshanarayana** [comment]
@vladimirvivien  @cpanato  PTAL. I have reorganised all the providers under `third_party` as suggested in the comments, and added backward compatibility type alias to make sure we do not break existing behaviours for the users who are already consuming this in some form.

**Comment by cpanato** [comment]
@harshanarayana rebase after we merge https://github.com/kubernetes-sigs/e2e-framework/pull/480

**Review by cpanato — APPROVED** [review]
/lgtm
